### DISCUSSION TITLE
HomeSpan and ESP-Now

### DISCUSSION THREAD

**Discussion by spyder66** [HomeSpan/HomeSpan#308] [General]
Does anyone have experience of getting HomeSpan and ESP-Now working together on the same ESP32 device.  Is this even possible?

I have a sketch using ESP-Now (master/slave) that works, and another one that is HomeSpan only, but as soon as I combine both features, ESP-Now stops sending data.  I presume it is because of some conflict with both HomeSpan and ESP-Now using Wifi.

**Comment by MnemonicFish** [comment]
I'm not saying it's not impossible but over coming the conflict became too much of an effort, the solution I found was to run a function in homespan to pass data via serial to a 2nd ESP broadcasting ESPNOW.

**Reply by spyder66** [reply]
Good solution.  I'd very much like to get this to work though.  Can anyone tell me the technical reasons as to why it doesn't work.  My Wifi experience is not the great, but I'll dig through the code anyways and try to understand how each piece works.  Any pointers would be greatly appreciated.

**Comment by MnemonicFish** [comment]
came across a new article;  it can be done by configuring the ESP32 as a wi-fi station and access point at the same time, you'd need to use different wi-fi channels, one for ESP-NOW and other to connect to Wi-Fi.  This article uses a webserver sensor dashboard + multiple ESPNOW boards;  https://randomnerdtutorials.com/esp32-esp-now-wi-fi-web-server/  

let us know how you get on integrating it into Homespan!

**Reply by spyder66** [reply]
Cool, thanks.  I will look into that and let you know how I get on.

**Reply by spyder66** [reply]
Had a go at implementing this. Got to the point where it would send a single message which was received by the receiver, but then all subsequent messages failed. If I reboot just the sender, same thing would happen again, only the first packet received. I'm assuming this points to an issue on the sending side, but haven't had time to investigate more yet.

**Reply by spyder66** [reply]
Ok, I have managed to get this working with a single ESP-NOW device that is also a HomeBridge device that can transmit data to 1 (or more) ESP-NOW slave devices.  I can now control the slave devices via the single HomeBridge device.

**Reply by MnemonicFish** [reply]
Great stuff 👍🏽 is HomeSpan running as expected with multiple ESPNOW slaves?

**Reply by blaisebarrette** [reply]
I find this project very interesting! I plan eventually on controlling the curtains. what I don’t like is the power supply and routing wires to each window. If each unit could be battery powered, that would be so much cleaner. ESP-NOW is supposed to be very power efficient. So a wired esp32 running Homespan while being a ESP-NOW master controlling each window as ESP-NOW slaves would be perfect in this scenario.

**Comment by HomeSpan** [comment]
I've not used ESP-NOW -  is it suitable for low-power use?   One of the drawbacks of HomeKit is that devices need constant connections to WiFi and batteries drain too fast.  I appreciate that ESP-NOW still may use the WiFi radio, but can a standalone ESP32 using just ESP-NOW run for extended periods of time on just battery power?  If so, this could be a good addition for future versions of HomeSpan.

**Reply by MnemonicFish** [reply]
ESPNOW protocol is similar to low-power 2.4GHz wireless connectivity, using ESP32 MAC address. The pairing between devices is needed prior to their communication, after pairing is done, connections are peer-to-peer, with no handshake required. If the board loses power or resets it will automatically reconnect,  thus allowing us to take advantage of Deep sleep mode with auto reconnect on wake up. 

Deep sleep mode the CPU and Wifi /RF are powered OFF, but the RTC memory and RTC peripherals are ON, enabling wake up via RTC_timer or RTC_GPIO pins, power consumption in Deep sleep is approximately 250micro/A according to Espressif datasheet, the protocol is pretty flexible supporting multiple configurations;  one-way point to point communication, one Master multiple slaves, one slave multiple Masters and two way communication point to point, so various power configurations can be utilised, i.e.  Master boards could be powered via 5V adapter and Slave boards battery powered set in Deep sleep mode.

**Reply by ssstraub** [reply]
> can a standalone ESP32 using just ESP-NOW run for extended periods of time on just battery power? If so, this could be a good addition for future versions of HomeSpan.

Yes, my understanding is that is a primary use case of ESP-NOW, though I haven't tried it yet myself. HomeSpan is great and if it's possitlbe to combine it's interfacing with HomeKit in a battery-powered package... LOOK OUT, WORLD! lol
